### ISSUE TITLE
Vite v5 / v6 gives CJS deprecation warning with Component testing

### ISSUE THREAD

**Issue by MikeMcC399** [cypress-io/cypress#29557]
- follow on from https://github.com/cypress-io/cypress/issues/28347

### Current behavior

Cypress Changelog [13.13.0](https://docs.cypress.io/guides/references/changelog#13-10-0) states "Added support for vite v5 to `@cypress/vite-dev-server`."

Running Component Testing with Cypress  and vite v5 or vite v6 results in the message:

> The CJS build of Vite's Node API is deprecated.

### Desired behavior

Running Component Testing with Cypress and vite v5 or v6 should not show any deprecation message related to Vite's Node API.

### Test code to reproduce

See https://github.com/cypress-io/github-action/tree/master/examples/component-tests

To run, use the following:

```shell
git clone https://github.com/cypress-io/github-action
cd github-action
git clean -xfd # if repeating
cd examples/component-tests
npm ci
npm test
```

For debug info, repeat with:

```shell
export VITE_CJS_TRACE=true
npm test
```

See also https://github.com/cypress-io/cypress/issues/29557#issuecomment-2180513085 for steps to reproduce without using an additional repository using `npm/vite-dev-server` directly.

### Cypress Version

First reported on version `13.10.0`
Still reproducible on version `14.1.0`

### Node version

~~`v20.13.1`~~ `v20.18.1` (issue is sensitive to Node.js version)
For instance, it does **not** reproduce on Node.js `v22.14.0`
GitHub Actions are locked to using `node20` at this time. GitHub Actions does not currently offer any higher Node.js for JavaScript Actions to run in.

### Operating System

~~Ubuntu `22.04.4` LTS~~
Ubuntu `22.04.5` LTS and `24.04.1` LTS

### Debug Logs

```text
$ export VITE_CJS_TRACE=true
npm test

> example-component-tests@2.0.0 test
> cypress run --component


DevTools listening on ws://127.0.0.1:41551/devtools/browser/d676f1ba-6901-4f0c-a113-4d46551aac70
Trace: The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.
    at warnCjsUsage (/home/mike/github/temp/github-action/examples/component-tests/node_modules/vite/index.cjs:33:3)
    at Object.<anonymous> (/home/mike/github/temp/github-action/examples/component-tests/node_modules/vite/index.cjs:3:1)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at require (node:internal/modules/helpers:179:18)
    at getVite (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/node_modules/@cypress/vite-dev-server/dist/getVite.js:14:16)
    at devServer (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/node_modules/@cypress/vite-dev-server/dist/devServer.js:13:40)
    at Object.handler (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_require_async_child.js:166:24)
    at invoke (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:185:18)
    at /home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:59:14
    at tryCatcher (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/bluebird/js/release/util.js:16:23)
    at Promise.attempt.Promise.try (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/bluebird/js/release/method.js:39:29)
    at Object.wrapChildPromise (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:58:23)
    at Object.wrap (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/dev-server.js:18:8)
    at RunPlugins.execute (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:155:26)
    at EventEmitter.<anonymous> (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/child/run_plugins.js:56:12)
    at EventEmitter.emit (node:events:519:28)
    at process.<anonymous> (/home/mike/.cache/Cypress/13.10.0/Cypress/resources/app/node_modules/@packages/server/lib/plugins/util.js:33:22)
    at process.emit (node:events:519:28)

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        13.10.0                                                                        │
  │ Browser:        Electron 118 (headless)                                                        │
  │ Node Version:   v20.13.1 (/home/mike/n/bin/node)                                               │
  │ Specs:          1 found (Stepper.cy.jsx)                                                       │
  │ Searched:       **/*.cy.{js,jsx,ts,tsx}                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────

  Running:  Stepper.cy.jsx                                                                  (1 of 1)


  <Stepper />
    ✓ renders (60ms)
    ✓ stepper should default to 0 (36ms)
    ✓ supports a "count" prop to set the value (26ms)
    ✓ when the increment button is pressed, the counter is incremented (116ms)
    ✓ when the decrement button is pressed, the counter is decremented (111ms)
    ✓ clicking + fires a change event with the incremented value (110ms)


  6 passing (530ms)


  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        6                                                                                │
  │ Passing:      6                                                                                │
  │ Failing:      0                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  0                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     0 seconds                                                                        │
  │ Spec Ran:     Stepper.cy.jsx                                                                   │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  Stepper.cy.jsx                           531ms        6        6        -        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        531ms        6        6        -        -        -
```

#### Installed npm module versions

```text
├── @types/react-dom@18.2.13
├── @types/react@18.2.28
├── @vitejs/plugin-react@4.2.1
├── cypress@13.10.0
├── react-dom@18.2.0
├── react@18.2.0
└── vite@5.2.11
```

**Comment by codyjohnsonfb** [comment]
Related to: https://github.com/cypress-io/cypress/issues/28373
Also still seeing this warning when running scripts.

**Comment by MikeMcC399** [comment]
@codyjohnsonfb 

- Thanks for linking in #28373! I had missed that issue (even though I commented on it quite a while ago!) 🤭 I guess it is the same root cause. Anyway my new issue doesn't have any TypeScript dependency in it, so I will leave it open in parallel to your issue.

**Comment by MikeMcC399** [comment]
## Steps to reproduce

```shell
git clone https://github.com/cypress-io/cypress
cd cypress
n auto
npm install yarn@latest -g
yarn
cd npm/vite-dev-server
yarn test
```

## Logs

```text
yarn test
yarn run v1.22.22
$ yarn test-unit
$ mocha -r ts-node/register/transpile-only --config ./test/.mocharc.js
The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.

```

<details>
  <summary>Full Log</summary>

```test
yarn test
yarn run v1.22.22
$ yarn test-unit
$ mocha -r ts-node/register/transpile-only --config ./test/.mocharc.js
The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.


  initCypressTests
    support file / spec file loading
      ✔ doesn't load the support file if one is not provided
      ✔ load the support file along with the spec
      empty devServerPublicPathRoute
        ✔ load the support file along with the spec
      windows
        ✔ doesn't load the support file if one is not provided
        ✔ load the support file along with the spec
        empty devServerPublicPathRoute
          ✔ load the support file along with the spec

  resolveConfig
    config resolution: v4
📦 Found package.json for project vite4-inspect.
📦 Creating a new node_modules cache dir at /tmp/cy-system-tests-node-modules/vite4-inspect/node_modules
📦 node_modules symlink created at /tmp/cy-projects/vite4-inspect/node_modules
/tmp/cy-projects/vite4-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884953398" in /tmp/cy-projects/vite4-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
warning Ignored scripts due to flag.
📦 Copying yarn.lock and fixing relative paths for vite4-inspect
      ✔ with <project-root>/vite.config.js (17001ms)
📦 Found package.json for project vite4-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite4-inspect/node_modules
/tmp/cy-projects/vite4-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884970377" in /tmp/cy-projects/vite4-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite4-inspect
      ✔ with component.devServer.viteConfig provided (300ms)
📦 Found package.json for project vite4-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite4-inspect/node_modules
/tmp/cy-projects/vite4-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884970698" in /tmp/cy-projects/vite4-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite4-inspect
      ✔ calls viteConfig if it is a function (298ms)
    file watching
📦 Found package.json for project vite4-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite4-inspect/node_modules
/tmp/cy-projects/vite4-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884971005" in /tmp/cy-projects/vite4-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite4-inspect
      ✔ should be disabled in run mode
📦 Found package.json for project vite4-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite4-inspect/node_modules
/tmp/cy-projects/vite4-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884971309" in /tmp/cy-projects/vite4-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite4-inspect
      ✔ uses defaults in open mode
    config resolution: v5
📦 Found package.json for project vite5-inspect.
📦 Creating a new node_modules cache dir at /tmp/cy-system-tests-node-modules/vite5-inspect/node_modules
📦 node_modules symlink created at /tmp/cy-projects/vite5-inspect/node_modules
/tmp/cy-projects/vite5-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884971623" in /tmp/cy-projects/vite5-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
warning Ignored scripts due to flag.
📦 Copying yarn.lock and fixing relative paths for vite5-inspect
      ✔ with <project-root>/vite.config.js (25007ms)
📦 Found package.json for project vite5-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite5-inspect/node_modules
/tmp/cy-projects/vite5-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884996798" in /tmp/cy-projects/vite5-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite5-inspect
      ✔ with component.devServer.viteConfig provided (1079ms)
📦 Found package.json for project vite5-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite5-inspect/node_modules
/tmp/cy-projects/vite5-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884997883" in /tmp/cy-projects/vite5-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite5-inspect
      ✔ calls viteConfig if it is a function (602ms)
    file watching
📦 Found package.json for project vite5-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite5-inspect/node_modules
/tmp/cy-projects/vite5-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884998750" in /tmp/cy-projects/vite5-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite5-inspect
      ✔ should be disabled in run mode
📦 Found package.json for project vite5-inspect.
📦 node_modules symlink created at /tmp/cy-projects/vite5-inspect/node_modules
/tmp/cy-projects/vite5-inspect/yarn.lock
📦 Writing yarn.lock with fixed relative paths to temp dir
📦 Running "yarn install --prefer-offline --ignore-scripts --frozen-lockfile --update-checksums --ignore-engines --cache-folder=/tmp/cy-system-tests-yarn-cache/1718884999183" in /tmp/cy-projects/vite5-inspect
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
success Already up-to-date.
📦 Copying yarn.lock and fixing relative paths for vite5-inspect
      ✔ uses defaults in open mode


  16 passing (46s)

Done in 49.42s.
```
</details>

**Comment by MikeMcC399** [comment]
This issue is also reproducible with` vite@6.0.9` and `cypress@14.0.0`

```text
DevTools listening on ws://127.0.0.1:36773/devtools/browser/d648be15-2f7b-4319-8809-9b987df35ca7
The CJS build of Vite's Node API is deprecated. See https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        14.0.0                                                                         │
  │ Browser:        Electron 130 (headless)                                                        │
  │ Node Version:   v20.18.0 (/__e/node20/bin/node)                                                │
  │ Specs:          1 found (Stepper.cy.jsx)                                                       │
  │ Searched:       **/*.cy.{js,jsx,ts,tsx}                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Edit: The outcome depends on the version of Node.js

| Node.js  | reproducible |
| -------- | ------------ |
| v18.20.6 | yes          |
| v20.18.2 | yes          |
| v22.13.1 | no           |
| v23.6.1  | no           |

**Comment by MikeMcC399** [comment]
The issue is no longer reproducible in the workflow https://github.com/cypress-io/github-action/actions/workflows/example-component-test.yml with Cypress `14.2.1`. GitHub Actions has migrated `node20` to use [Node.js v20.19.0 LTS](https://nodejs.org/en/blog/release/v20.19.0) which includes backported `require(esm)` functionality. Presumably this now masks the issue which remains reproducible locally under the older 20.x release [Node.js v20.18.3](https://nodejs.org/en/blog/release/v20.18.3).

Results testing locally:

| Node.js release line | Latest     | Deprecation warning |
| -------------------- | ---------- | ------------------- |
| `18.x`               | `v18.20.8` | YES                 |
| `20.x`               | `v20.19.0` | NO                  |
| `22.x`               | `v22.14.0` | NO                  |
| `23.x`               | `v23.10.0` | NO                  |

This issue can be closed after Node.js `18.x` enters [end-of-life](https://github.com/nodejs/release#release-schedule) (planned Apr 30, 2025) since other remaining supported Node.js versions no longer exhibit the issue.

**Comment by MikeMcC399** [comment]
Each latest version of the supported Node.js release lines shows no deprecation warning for Vite v6:

| Node.js release line | Latest     | Deprecation warning |
| -------------------- | ---------- | ------------------- |
| `20.x`               | `v20.19.1` | NO                  |
| `22.x`               | `v22.15.0` | NO                  |
| `23.x`               | `v23.11.0` | NO                  |
| `24.x`               | `v24.0.1`  | NO                  |

This issue is therefore resolved through enhancements in Node.js and can be closed.

**Comment by MikeMcC399** [comment]
Re-opening: the issue is still reproducible for Vite v5 and according to https://docs.cypress.io/app/component-testing/get-started#Supported-Frameworks versions Vite 4-6 are supported by Cypress.

Ubuntu `22.04.2` LTS, Node.js `22.15.0` LTS, Vite `5.4.19`

```shell
git clone --branch test/vite-v5-ct https://github.com/MikeMcC399/github-action
cd github-action/examples/component-tests
npm install vite@5 -D
npm test
```

```text
$ npm test

> example-component-tests@2.0.0 test
> cypress run --component


DevTools listening on ws://127.0.0.1:35113/devtools/browser/8f91a36e-4167-4a8a-a369-ae4537d2f29c
The CJS build of Vite's Node API is deprecated. See https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        14.3.3                                                                         │
  │ Browser:        Electron 130 (headless)                                                        │
  │ Node Version:   v22.15.0 (/home/mike/n/bin/node)                                               │
  │ Specs:          1 found (Stepper.cy.jsx)                                                       │
  │ Searched:       **/*.cy.{js,jsx,ts,tsx}                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```

Possibly this deprecation warning in Vite v5 can simply be ignored if Node.js versions can work without a migration.

https://vite.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated

**Comment by AtofStryker** [comment]
Looks like Vite 7 is removing the cjs build so work is going to need to be done to this package to cut it over to ESM only. See https://github.com/vitejs/vite/blob/v7.0.0-beta.0/packages/vite/CHANGELOG.md#700-beta0-2025-06-02

**Comment by tobiashm** [comment]
Vite 7 has been out for a couple of weeks now, and it does indeed break with Cypress.


> Could not find "vite" in your project's dependencies. Please install "vite" to fix this error.
> 
> Error [ERR_REQUIRE_ESM]: require() of ES Module /Users/tobias/test-project/node_modules/vite/dist/node/index.js from /Users/tobias/Library/Caches/Cypress/14.5.2/Cypress.app/Contents/Resources/app/packages/server/node_modules/@cypress/vite-dev-server/dist/getVite.js not supported. Instead change the require of index.js in /Users/tobias/Library/Caches/Cypress/14.5.2/Cypress.app/Contents/Resources/app/packages/server/node_modules/@cypress/vite-dev-server/dist/getVite.js to a dynamic import() which is available in all CommonJS modules.

**Comment by MikeMcC399** [comment]
@tobiashm 

> Vite 7 has been out for a couple of weeks now, and it does indeed break with Cypress.

Which Node.js version are you using? 

It's running successfully for E2E testing in https://github.com/cypress-io/github-action/tree/master/examples/wait-on-vite see logs https://github.com/cypress-io/github-action/actions/runs/16320830830/job/46097867350

**Comment by MikeMcC399** [comment]
@tobiashm 

> Vite 7 has been out for a couple of weeks now, and it does indeed break with Cypress.

This issue is about Cypress Component Testing with Vite v5 and v6 

For Vite v7 see https://github.com/cypress-io/cypress/issues/31882#issuecomment-3074056521 which says:

> - Vite 7 CT support is going to be available in Cypress 15 with https://github.com/cypress-io/cypress/pull/32001. I am going to close this issue as complete.

**Comment by AtofStryker** [comment]
closing as fixed with https://github.com/cypress-io/cypress/pull/32001 and will be available in Cypress 15
